### PR TITLE
Add tests for `GenericTypeRegistry`

### PR DESCRIPTION
**Pull request by amomchilov**
### Motivation

I wrote these unit tests for myself while exploring #1634. They seem worth keeping around.

### CODE REVIEW

File: spec/tapioca/runtime/generic_type_registry_spec.rb [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+          end
+        end
+
+        describe ".lookup_type_variables" do
+          it "returns nil for non-generic types" do
+            assert_nil(GenericTypeRegistry.lookup_type_variables(Object))
+          end
+
+          it "returns the type variables for generic types" do
+            assert_equal([SampleGenericClass::Element], GenericTypeRegistry.lookup_type_variables(SampleGenericClass))
```
Comment (Morriar) [inline spec/tapioca/runtime/generic_type_registry_spec.rb:26]: Could we nest the test class in the test itself with a `sample_class = Class.new do ...`?

File: spec/tapioca/runtime/generic_type_registry_spec.rb [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+            assert_same(fake_type_member1, type_variables[0])
+            assert_same(fake_type_member2, type_variables[1])
+          end
+        end
+
+        describe "the patch for .inherited on generic classes" do
+          # This is more of an internal detail and cross-cutting concern of all the public APIs,
+          # but it's easier to test here on its own.
+
+          it "rescues exceptions that would have prevents subclassing" do
```
Comment (Morriar) [inline spec/tapioca/runtime/generic_type_registry_spec.rb:70]:
```suggestion
          it "rescues exceptions that would prevent subclassing" do
```

File: spec/tapioca/runtime/generic_type_registry_spec.rb [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+            assert_raises(RuntimeError) do
+              # Precondition. If not caught by the patch, this error would have prevented our subclassing.
+              RaisesInInheritedCallback.inherited(Class.new)
```
Comment (Morriar) [inline spec/tapioca/runtime/generic_type_registry_spec.rb:73]: Same thing here, can we nest the test class inside the test?

Comment (amomchilov) [reply spec/tapioca/runtime/generic_type_registry_spec.rb:73]: Oh yeah, this one only exists for this test case. Done!

Comment (amomchilov) [reply spec/tapioca/runtime/generic_type_registry_spec.rb]: Ah wait, Sorbet doesn't like this, and I'm not sure how to convince it haha

Here's my attempt: https://github.com/Shopify/tapioca/commit/838d4ffefa07e9f4582e4732dd6c2dbf01575276

```
spec/tapioca/runtime/generic_type_registry_spec.rb:74: Method type_member does not exist on T::Class[Object] https://srb.help/7003
    74 |              Element = type_member
                                ^^^^^^^^^^^
  Got T::Class[Object] originating from:
    spec/tapioca/runtime/generic_type_registry_spec.rb:71:
    71 |            cls = Class.new do
    72 |              extend T::Generic
    73 |
    74 |              Element = type_member
    75 |
       |...
    78 |                  super
    79 |                  raise "Boom"
    80 |                end
    81 |              end
    82 |            end

spec/tapioca/runtime/generic_type_registry_spec.rb:89: Method [] does not exist on T::Class[Object] https://srb.help/7003
    89 |            result = cls[Object] # Should be a distinct subclass
                             ^^^^^^^^^^^
  Got T::Class[Object] originating from:
    spec/tapioca/runtime/generic_type_registry_spec.rb:71:
    71 |            cls = Class.new do
    72 |              extend T::Generic
    73 |
    74 |              Element = type_member
    75 |
       |...
    78 |                  super
    79 |                  raise "Boom"
    80 |                end
    81 |              end
    82 |            end
Errors: 2
```

**Review by Morriar — APPROVED** [review]

File: spec/tapioca/runtime/generic_type_registry_spec.rb (hunk 1 above)
Comment (amomchilov) [reply spec/tapioca/runtime/generic_type_registry_spec.rb:26]: https://github.com/Shopify/tapioca/pull/1635#discussion_r1313683273

**Pull request merged**